clients/drcachesim/tests/record_filter_unit_tests.cpp

```diff
@@ -94,7 +126,122 @@ get_basic_counts(const std::string &trace_dir)
+        { { TRACE_TYPE_WRITE, 4, { 10 } }, { true, false } },
+        { { TRACE_TYPE_INSTR, 4, { 11 } }, { true, false } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_VIRTUAL_ADDRESS, { 12 } },
+          { true, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_PHYSICAL_ADDRESS, { 13 } },
+          { true, true } },
+        { { TRACE_TYPE_READ, 4, { 14 } }, { true, false } },
+        { { TRACE_TYPE_INSTR, 4, { 15 } }, { true, false } },
+        /* Unit header. */
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 16 } }, { true, false } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 17 } }, { true, false } },
+        /* Unit header. */
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 18 } }, { true, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 19 } }, { true, true } },
+        { { TRACE_TYPE_INSTR, 4, { 20 } }, { true, false } },
+        /* First half supposed to end here. See SPLIT_AT_INSTR_COUNT. */
+        { { TRACE_TYPE_INSTR, 4, { 21 } }, { false, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_VIRTUAL_ADDRESS, { 22 } },
+          { false, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_PHYSICAL_ADDRESS_NOT_AVAILABLE, { 23 } },
+          { false, true } },
+        { { TRACE_TYPE_READ, 4, { 24 } }, { false, true } },
+        { { TRACE_TYPE_WRITE, 4, { 25 } }, { false, true } },
+        /* Unit header. */
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 26 } }, { false, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 27 } }, { false, true } },
+        { { TRACE_TYPE_INSTR, 4, { 28 } }, { false, true } },
+        { { TRACE_TYPE_READ, 4, { 29 } }, { false, true } },
+        { { TRACE_TYPE_WRITE, 4, { 30 } }, { false, true } },
+        /* Trace shard footer. */
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CHUNK_FOOTER, { 31 } }, { true, true } },
+        { { TRACE_TYPE_FOOTER, 0, { 32 } }, { true, true } }
+    };
+
+    /* Check each half. */
+    for (int k = 0; k < 2; ++k) {
+        dynamorio::drmemtrace::record_filter_t::record_filter_func_t *toggle_filter =
+            new dynamorio::drmemtrace::toggle_filter_t(SPLIT_AT_INSTR_COUNT, k == 0);
+        test_record_filter_t *record_filter = new test_record_filter_t({ toggle_filter });
+        void *shard_data = record_filter->parallel_shard_init_stream(0, nullptr, nullptr);
+        for (int i = 0; i < (int)entries.size(); ++i) {
+            if (!record_filter->parallel_shard_memref(shard_data, entries[i].entry)) {
+                fprintf(stderr, "Filtering failed\n");
+                return false;
+            }
+        }
+        if (!record_filter->parallel_shard_exit(shard_data)) {
+            fprintf(stderr, "Filtering exit failed\n");
+            return false;
+        }
+        std::vector<trace_entry_t> half = record_filter->get_output_entries();
+        delete record_filter;
+        delete toggle_filter;
```

Comment:
Better to use unique_ptr and pass to the vector with get() and avoid needing delete

Reply:
Done.